numpy/_array_api/_creation_functions.py

```diff
@@ -0,0 +1,177 @@
+from __future__ import annotations
+
+from ._types import (Optional, SupportsDLPack, SupportsBufferProtocol, Tuple,
+                     Union, array, device, dtype)
+from ._dtypes import _all_dtypes
+
+import numpy as np
+
+def asarray(obj: Union[float, NestedSequence[bool|int|float], SupportsDLPack, SupportsBufferProtocol], /, *, dtype: Optional[dtype] = None, device: Optional[device] = None, copy: Optional[bool] = None) -> array:
+    """
+    Array API compatible wrapper for :py:func:`np.asarray <numpy.asarray>`.
+
+    See its docstring for more information.
+    """
+    # _array_object imports in this file are inside the functions to avoid
+    # circular imports
+    from ._array_object import ndarray
+    if device is not None:
+        # Note: Device support is not yet implemented on ndarray
+        raise NotImplementedError("Device support is not yet implemented")
+    if copy is not None:
+        # Note: copy is not yet implemented in np.asarray
+        raise NotImplementedError("The copy keyword argument to asarray is not yet implemented")
+    if isinstance(obj, ndarray):
+        return obj
+    res = np.asarray(obj, dtype=dtype)
+    if res.dtype not in _all_dtypes:
+        raise TypeError(f"The array_api namespace does not support the dtype '{res.dtype}'")
+    return ndarray._new(res)
+
+def arange(start: Union[int, float], /, *, stop: Optional[Union[int, float]] = None, step: Union[int, float] = 1, dtype: Optional[dtype] = None, device: Optional[device] = None) -> array:
+    """
+    Array API compatible wrapper for :py:func:`np.arange <numpy.arange>`.
+
+    See its docstring for more information.
+    """
+    from ._array_object import ndarray
+    if device is not None:
+        # Note: Device support is not yet implemented on ndarray
+        raise NotImplementedError("Device support is not yet implemented")
+    return ndarray._new(np.arange(start, stop=stop, step=step, dtype=dtype))
+
+def empty(shape: Union[int, Tuple[int, ...]], /, *, dtype: Optional[dtype] = None, device: Optional[device] = None) -> array:
+    """
+    Array API compatible wrapper for :py:func:`np.empty <numpy.empty>`.
+
+    See its docstring for more information.
+    """
+    from ._array_object import ndarray
+    if device is not None:
+        # Note: Device support is not yet implemented on ndarray
+        raise NotImplementedError("Device support is not yet implemented")
+    return ndarray._new(np.empty(shape, dtype=dtype))
+
+def empty_like(x: array, /, *, dtype: Optional[dtype] = None, device: Optional[device] = None) -> array:
```

Comment:
FYI the positional-only specifier of [PEP 570](https://www.python.org/dev/peps/pep-0570/) requires python >= 3.8,
so this will crash for python 3.7.

Reply:
Yes, that's on purpose. Reason: `empty_like(x)` is portable, `empty_like(x=x)` is not because other libraries tend to use other names for positional parameters (e.g. PyTorch consistently uses `input`). For variables like `x` that are not meaningful or commonly used as named parameters, the choice was therefore made to make them positional-only.

There's a couple of places where those annotations need a tweak, there's on open issue about that. But the principle above should hold. `x=`, `y=` should be made positional-only, while `dtype=` or `shape=` clearly shouldn't.